kingpin/actors/rightscale/server_array.py

```diff
@@ -421,6 +421,187 @@ def _execute(self):
+        if ('image_href' in params and params['image_href'] == 'default'):
+            params['image_href'] = yield self._find_def_image_href(instance)
+
+        # Second pass at the generating the parameters. We did this at
+        # instantiation time as a sanity check to make sure the parameters were
+        # half-decent. Now we run it a second time in case any the 'image_href'
+        # magic above was executed.
+        rs_params = self._generate_rightscale_params(
+            'instance', self.option('params'))
+
+        self.log.info('Updating array\'s next_instance "%s" with params: %s' %
+                      (instance.soul['name'], rs_params))
+        try:
+            yield self._client.update(instance, rs_params)
+        except requests.exceptions.HTTPError as e:
+            if e.response.status_code in (422, 400):
+                msg = ('Invalid parameters supplied to patch array "%s"' %
+                       self.option('array'))
+                raise exceptions.RecoverableActorFailure(msg)
+            raise
+
+        raise gen.Return()
+
+    @gen.coroutine
+    def _find_def_image_href(self, instance):
+        self.log.debug('Searching for default boot AMI for %s' %
+                       instance.soul['name'])
+
+        # Find the MultiCloudImage associated with this 'instance' object, then
+        # get the full list of 'settings' for that MCI.
+        mci = yield self._client.show(instance.multi_cloud_image)
+        self.log.debug('Got MCI: %s' % mci.soul['name'])
+        mci_settings = yield self._client.show(mci.settings)
+        self.log.debug('Got %s MCI Cloud Settings.' % len(mci_settings))
+
+        # Now, find the 'setting' that matches the cloud of our instance. Note,
+        # there should never be more than one returned -- so we take the first
+        # one in the list and save it.
+        try:
+            setting = [s for s in mci_settings if
+                       s.cloud.path == instance.cloud.path][0]
+            image_href = [l['href'] for l in setting.soul['links']
+                          if l['rel'] == 'image'][0]
+        except KeyError:
+            raise InvalidInputs(
+                'Unable to locate default image_href for %s.' % instance.soul)
+
+        self.log.info('%s default AMI HREF found: %s' %
+                      (instance.soul['name'], image_href))
+
+        raise gen.Return(image_href)
+
+    @gen.coroutine
+    def _execute(self):
+        # First, find the arrays we're going to be patching.
+        arrays = yield self._find_server_arrays(
+            self.option('array'), exact=self.option('exact'))
+
+        # In dry run, just comment that we would have made the change.
+        if self._dry:
+            self.log.debug('Not making any changes.')
+            self.log.info('Params would be: %s' % self.option('params'))
+            raise gen.Return()
+
+        # Do the real work
+        yield self._apply(self._update_params, arrays)
```

Comment:
What's `_apply`? Also I think we've discussed in person that _execute should do much of the heavy lifting rather than calling another function that does everything